### PR TITLE
Implement new Discovery Pane API (fix #2447)

### LINKED ISSUE #2447: Implement API for disco pane
We need a v3 api endpoint for `addons-frontend` to get the list of addons for the new disco pane.

Some data that we need:

 * icon
 * some sort of slogan (see invision screen, "Block ads with ...")
 * description
 * name
 * addon-id

Do we something specific about themes?

### PR DESCRIPTION
**Pull request by diox**

### CODE REVIEW

File: docs/topics/api/discovery.rst [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+This endpoint allows you to fetch content for the new Discovery Pane in
+Firefox (about:addons).
+
+ .. http:get:: /api/v3/discovery/
+
+    :>json int count: The number of results for this query.
+    :>json array results: An array of :ref:`discovery items <discovery-item>`.
+    :>json string results[].heading: The heading for this item. If present, it contains the following sub-strings, that the client needs to use to format the string as it desires: ``{start_sub_heading}``, ``{end_sub_heading}`` and ``{addon_name}``.
+    :>json string|null results[].description: The description for this item, if any. May contain some HTML tags.
+    :>json object results[].addon: The :ref:`add-on <addon-detail-object>` for this item. Only a subset of fields are present: ``id``, ``current_version``, ``icon_url``, ``theme_data``, ``type`` and ``url``.
```
Comment (mstriemer) [inline docs/topics/api/discovery.rst]: Is `id` the guid, and `url` the install URL? We need the guid and install URL to run the add-on manager API.

Comment (mstriemer) [inline docs/topics/api/discovery.rst]: Actually I guess the install URL comes from `current_version.files`.

Comment (diox) [inline docs/topics/api/discovery.rst]: `url` is the detail URL, you're right, the `download_url` comes from `current_version.files`. I'll add the `guid` back.

File: src/olympia/discovery/serializers.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    description = serializers.CharField()
+    addon = DiscoveryAddonSerializer()
+
+    def to_representation(self, instance):
+        data = super(DiscoverySerializer, self).to_representation(instance)
+        if data['heading'] is None:
+            if instance.addon.listed_authors:
+                data['heading'] = u'%s by %s' % (
+                    unicode(instance.addon.name),
+                    instance.addon.listed_authors[0].name)
```
Comment (diox) [inline src/olympia/discovery/serializers.py]: I'll simplify the handling of the heading and remove this, because some themes actually include `by <artist>` in the theme name already, and it's annoying to detect with translations, so let's just drop it. Plus design copy didn't have it on the last theme anyway.

**Comment by diox** [comment]
I'll look at those test failures, they look super weird.

File: src/olympia/discovery/api_urls.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
+from django.conf.urls import include, url
+
+from rest_framework.routers import SimpleRouter
+
+from . import views
+
+# Because we're registering things at "/discovery/" we don't want the trailing
+# slash that comes with the Router by default.
```
Comment (EnTeQuAk) [inline src/olympia/discovery/api_urls.py]: Either I'm not following what you mean or... we don't?

File: src/olympia/discovery/data.py [hunk 4]
```diff
@@ -1,60 +1,62 @@
...
                   '{end_sub_heading}'),
         description=string_concat(
             '<blockquote>',
-            _('“Download videos in a single click.”'),
+            _(u'“Download videos in a single click.”'),
             '<cite>— Carpe Diem</cite>',
-            '</blockquote')),
+            '</blockquote>')),
+
+    DiscoItem(addon_id=686505),
```
Comment (EnTeQuAk) [inline src/olympia/discovery/data.py]: Thos are really valid entries?

File: src/olympia/discovery/serializers.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
+from rest_framework import serializers
+
+from olympia.addons.models import Addon
+from olympia.addons.serializers import AddonSerializer, VersionSerializer
+from olympia.versions.models import Version
+
+
+class DiscoveryVersionSerializer(VersionSerializer):
+    class Meta:
+        fields = ('compatibility', 'files',)
```
Comment (EnTeQuAk) [inline src/olympia/discovery/serializers.py]: you don't need the trailing comma here.

File: src/olympia/discovery/serializers.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+        model = Version
+
+
+class DiscoveryAddonSerializer(AddonSerializer):
+    current_version = DiscoveryVersionSerializer()
+
+    class Meta:
+        fields = ('id', 'current_version', 'icon_url',
+                  'theme_data', 'type', 'url',)
```
Comment (EnTeQuAk) [inline src/olympia/discovery/serializers.py]: same here, no trailing comma.

File: src/olympia/discovery/api_urls.py (hunk 3 above)
Comment (EnTeQuAk) [inline src/olympia/discovery/api_urls.py]: Ah, because it would result in `/v1/discovery//`, gotcha.

But we really miss the trailing slashes in case you do a sub-route, e.g

```python
discovery.register(r'thunderbird', views.ThunderbirdDiscoveryViewSet, base_name='discovery-thunderbird')
```

'cause it's a `prefix` `r'thunderbird/` should not work imho.

Imho it's best to include the urlpatterns here in the root like this:

```python
url(r'^v3/', include('olympia.discovery.api_urls')),
```

and then go ahead and define sub-routes in the discovery specific urls.py

```python
router = DefaultRouter()
router.register('discovery', views.DiscoveryViewSet, 'discovery')
```

And, if we need that, let's use drf-extensions: https://chibisov.github.io/drf-extensions/docs/#nested-routes

File: src/olympia/discovery/tests/test_views.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+            result = response.data['results'][i]
+            assert result['addon']['id'] == item.addon_id
+            if item.heading:
+                assert result['heading'] == item.heading
+            else:
+                assert result['heading'] == u'%s by %s' % (
+                    unicode(addons[item.addon_id].name),
+                    addons[item.addon_id].listed_authors[0].name)
+            assert result['description'] == item.description
+            assert result['current_version']
```
Comment (EnTeQuAk) [inline src/olympia/discovery/tests/test_views.py]: I really do hate asserts nested in for loops because they're hard to debug / reason about if they fail. Wouldn't it be better anyway to overwrite the original `discopane_items` list and have a handful ones that we're testing here and not do a for loop?

File: src/olympia/discovery/api_urls.py (hunk 3 above)
Comment (diox) [inline src/olympia/discovery/api_urls.py]: Yeah, I don't like doing the `v3/` include because then it may overwrite anything, but you're right, that's more flexible, I'll change it.

File: src/olympia/discovery/data.py (hunk 4 above)
Comment (diox) [inline src/olympia/discovery/data.py]: Yes, but those are themes, for which we don't give an editorialized title/description.

File: src/olympia/discovery/serializers.py (hunk 5 above)
Comment (diox) [inline src/olympia/discovery/serializers.py]: But but... I like it :) It's convenient when adding new fields.

File: src/olympia/discovery/tests/test_views.py (hunk 7 above)
Comment (diox) [inline src/olympia/discovery/tests/test_views.py]: Yeah I can do that.

File: src/olympia/landfill/generators.py [hunk 8]
```diff
@@ -40,7 +40,7 @@ def _yield_name_and_cat(num, app=None):
         yield (addon_name, cat)
 
 
-def create_addon(name, icon_type, application):
+def create_addon(name, icon_type, application, **extra_kwargs):
```
Comment (diox) [inline src/olympia/landfill/generators.py]: Bonus to be able to create addons using the landfill generator function while passing extra fields, like... `id`.

File: src/olympia/landfill/generators.py [hunk 9]
```diff
@@ -86,7 +87,7 @@ def generate_addons(num, owner, app_name):
         generate_ratings(addon, 5)
 
 
-def create_theme(name):
+def create_theme(name, **extra_kwargs):
```
Comment (diox) [inline src/olympia/landfill/generators.py]: Same as above but for themes.

**Comment by diox** [comment]
@EnTeQuAk should be ready for review

File: docs/conf.py [hunk 10]
```diff
@@ -177,19 +181,19 @@
...
 
-# Grouping the document tree into LaTeX files. List of tuples
-# (source start file, target name, title, author, documentclass [howto/manual]).
+# Grouping the document tree into LaTeX files. List of tuples (source start
+# file, target name, title, author, documentclass [howto/manual]).
 latex_documents = [
-  ('index', 'olympia.tex', u'olympia Documentation',
-   u'Jeff Balogh', 'manual'),
+    ('index', 'olympia.tex', u'olympia Documentation', u'Mozilla Addons Team',
+     'manual'),
```
Comment (EnTeQuAk) [inline docs/conf.py]: Not really part of this PR but I don't care enough :)

File: src/olympia/api/parsers.py [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+    """
+
+    def parse(self, stream, media_type=None, parser_context=None):
+        """
+        Parses the incoming bytestream as a multipart encoded form,
+        and returns a DataAndFiles object.
+
+        `.data` will be a `QueryDict` containing all the form parameters.
+        `.files` will be a `QueryDict` containing all the form files.
+        +
```
Comment (EnTeQuAk) [inline src/olympia/api/parsers.py]: those `+`es look weird.

File: src/olympia/api/parsers.py [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+        +        For POSTs, accept Django request parsing.  See issue #3951.
+        """
+        parser_context = parser_context or {}
+        request = parser_context['request']
+        _request = request._request
+        if _request.method == 'POST':
+            return DataAndFiles(_request.POST, _request.FILES)
+        return super(MultiPartParser, self).parse(
+            stream, media_type=media_type, parser_context=parser_context)
```
Comment (EnTeQuAk) [inline src/olympia/api/parsers.py]: Yiekes! I ran into this bug, that sucks :cry:

File: src/olympia/discovery/api_urls.py [hunk 13]
```diff
@@ -0,0 +1,12 @@
+from django.conf.urls import include, url
+
+from rest_framework.routers import SimpleRouter
+
+from . import views
+
+discovery = SimpleRouter()
+discovery.register('discovery', views.DiscoveryViewSet, base_name='discovery')
```
Comment (EnTeQuAk) [inline src/olympia/discovery/api_urls.py]: :+1:

File: src/olympia/discovery/data.py (hunk 4 above)
Comment (EnTeQuAk) [inline src/olympia/discovery/data.py]: I see. For some reason I'd love to be able to know that those are themes just by looking at it but let's just keep that in mind for future improvements. wfm for now

File: src/olympia/discovery/serializers.py (hunk 5 above)
Comment (EnTeQuAk) [inline src/olympia/discovery/serializers.py]: I only find it convenient if you have something like

```python
foo = (
    'bar',
    'foo',
)
```
But here. Well, I really don't care and we don't have any guidelines yet. By the way, let's have some more guidelines! :laughing:

**Comment by EnTeQuAk** [comment]
r+

File: src/olympia/discovery/data.py (hunk 4 above)
Comment (diox) [inline src/olympia/discovery/data.py]: Yeah, I initially had the `type` but it was completely useless since it was also present in the `addon` instance so I just removed it.
